util/persistence.py

```diff
@@ -0,0 +1,242 @@
+        self.load = load
+        self.inter_secs = inter_secs
+        self.keep_n_inters = keep_n_inters
+        self.keep_n_epochs = keep_n_epochs
+        self.init_op = init_op if init_op else tf.global_variables_initializer()
+        self.saver = saver if saver else tf.train.Saver()
+        self._t0 = None
+        if checkpoint_dir:
+            self._csv = os.path.join(checkpoint_dir, 'results.csv')
+            if os.path.isfile(self._csv):
+                # read all lines and columns from results.csv
+                self._results = [l.split(',') for l in open(self._csv, 'r').readlines()[1:]]
+                # post-process lines to required data types of columns epoch, loss, dev-loss
+                self._results = [(int(r[0]), float(r[1]), None if len(r[2].strip()) == 0 else float(r[2])) for r in self._results]
+                # ordered by epoch
+                self._results = sorted(self._results, key=lambda r: r[0])
+            else:
+                self._results = []
+
+    def _get_checkpoints(self):
+        '''
+        Scans checkpoint dir and returns a list of checkpoint tuples
+        ordered by their global index.
+        Each checkpoint tuple consists of the
+         - epoch index (1-based, 0 for intermediate checkpoints),
+         - global index and a
+         - set of filenames.
+        '''
+        # list and hash for collecting checkpoint tuples
+        checkpoints = []
+        checkpoints_lookup = {}
+        # scanning checkpoint dir for checkpoints...
+        for filename in glob.glob(self.checkpoint_dir + '/*'):
+            # only process filenames of the following form
+            found = re.search('(inter|epoch-[0-9]+)_([0-9]+)\\.ckpt.*', filename, re.IGNORECASE)
+            if found:
+                # extract the epoch index - 0 for intermediate checkpoints
+                epoch_index = found.group(1)
+                epoch_index = int(epoch_index[6:]) if epoch_index.startswith('epoch-') else 0
+                # extract the global index (underscore part)
+                global_index = int(found.group(2))
+                # for looking up checkpoint tuples within this loop
+                key = '%d-%d' % (epoch_index, global_index)
+                if key in checkpoints_lookup:
+                    # we already processed some file(s) for this checkpoint...
+                    _, _, filenames = checkpoints_lookup[key]
+                else:
+                    # creating a new checkpoint tuple
+                    filenames = []
+                    checkpoint = (epoch_index, global_index, filenames)
+                    checkpoints_lookup[key] = checkpoint
+                    checkpoints.append(checkpoint)
+                # add current filename to checkpoint's file-set
+                filenames.append(filename)
+        # return list of checkpoint tuples - ordered by their global index
+        return sorted(checkpoints, key=lambda cp: cp[1])
```

Comment:
Maybe worthwhile to add some debug / logging in this method ? At least being able to see what has been picked up in term of filenames, and what information we extracted from them.

Reply:
This is a getter routine that will potentially called more than one time at startup. Added logging to `start` method.